zmon_agent/discovery/kubernetes/cluster.py

```diff
@@ -188,27 +191,50 @@ def get_cluster_pods(kube_client, cluster_id, alias, environment, region, infras
             'pod_host_ip': obj['status'].get('hostIP', ''),
             'pod_node_name': obj['spec']['nodeName'],
 
-            'containers': {
-                c['name']: {
-                    'image': c['image'],
-                    'ready': container_statuses.get(c['name'], {}).get('ready', True),
-                    'restarts': container_statuses.get(c['name'], {}).get('restartCount', 0),
-                    'ports': [p['containerPort'] for p in c.get('ports', []) if 'containerPort' in p],
-                } for c in containers
-            },
-
             'pod_phase': obj['status'].get('phase'),
             'pod_initialized': conditions.get('Initialized', False),
             'pod_ready': conditions.get('Ready', True),
-            'pod_scheduled': conditions.get('PodScheduled', False),
+            'pod_scheduled': conditions.get('PodScheduled', False)
         }
 
-        entity = add_labels_to_entity(entity, obj['metadata'].get('labels', {}))
-        entity = add_labels_to_entity(entity, obj['metadata'].get('annotations', {}))
+        pod_entity = {
+            'id': 'pod-{}-{}[{}]'.format(pod.name, pod.namespace, cluster_id),
+            'type': POD_TYPE,
+            'containers': {}
+        }
 
-        entities.append(entity)
+        pod_entity.update(shared_properties)
+        pod_entity.update(pod_labels)
+        pod_entity.update(pod_annotations)
+
+        for container in containers:
+            container_name = container['name']
+            container_image = container['image']
+            container_ready = container_statuses.get(container['name'], {}).get('ready', False)
+            container_restarts = container_statuses.get(container['name'], {}).get('restartCount', 0)
+            container_ports = [p['containerPort'] for p in container.get('ports', []) if 'containerPort' in p]
+
+            container_entity = {
+                'id': 'container-{}-{}-{}[{}]'.format(pod.name, pod.namespace, container_name, cluster_id),
+                'type': POD_TYPE,
+                'container_name': container_name,
+                'container_image': container_image,
+                'container_ready': container_ready,
+                'container_restarts': container_restarts,
+                'container_ports': container_ports
+            }
+            pod_entity['containers'][container_name] = {
+                'image': container_image,
+                'ready': container_ready,
+                'restarts': container_restarts,
+                'ports': container_ports
+            }
 
-    return entities
+            container_entity.update(pod_labels)
+            container_entity.update(shared_properties)
+            yield container_entity
```

Comment:
I suggest we keep it consistent with our returns. Either ``yield`` on all functions or just return a list.

Reply:
I'd change it to `yield` everywhere, then. What do you think?